src/SimpleInjector/Internals/InstanceProducerVisualizer.cs

```diff
@@ -62,52 +64,47 @@ internal static string VisualizeIndentedObjectGraph(this InstanceProducer produc
                 string.Join(", ", visualizedDependencies));
         }
 
-        private static string VisualizeIndentedObjectGraph(this InstanceProducer producer, int indentingDepth,
-            HashSet<InstanceProducer> set)
+        private static void VisualizeIndentedObjectGraph(this InstanceProducer producer, int indentingDepth, bool last, HashSet<InstanceProducer> set, ObjectGraphBuilder objectGraphBuilder)
         {
-            var visualizedDependencies =
-                from relationship in producer.GetRelationships()
-                let dependency = relationship.Dependency
-                let subGraph = dependency.VisualizeIndentedObjectSubGraph(indentingDepth + 1, set)
-                select Environment.NewLine + subGraph;
-
-            return string.Format(CultureInfo.InvariantCulture, "{0}{1}({2})",
-                new string(' ', indentingDepth * 4),
-                producer.ImplementationType.ToFriendlyName(),
-                string.Join(",", visualizedDependencies));
+            objectGraphBuilder.BeginInstanceProducer(producer);
+
+            var dependencies = producer
+                                .GetRelationships()
+                                .Select(relationship => relationship.Dependency)
+                                .ToList();
```

Comment:
Awesome. All looks good. One last comment, before we can merge this back. The coding style used by Simple Injector is to always use an indentation of 4 spaces, and not try to aline code with the previous line. To adhere with that style (I think it's a StyleCop style, but this is also something that originates from Kevin Henny (and others)) the code above should be formatted as follows:

``` c#
            var dependencies = producer
                .GetRelationships()
                .Select(relationship => relationship.Dependency)
                .ToList();
```